pkg/mssqlcontainer/ingest/ingest.go

```diff
@@ -0,0 +1,135 @@
+package ingest
+
+import (
+	"fmt"
+	"github.com/microsoft/go-sqlcmd/internal/container"
+	"github.com/microsoft/go-sqlcmd/internal/uri"
+	"github.com/microsoft/go-sqlcmd/pkg/mssqlcontainer/ingest/extract"
+	"github.com/microsoft/go-sqlcmd/pkg/mssqlcontainer/ingest/location"
+	"github.com/microsoft/go-sqlcmd/pkg/mssqlcontainer/ingest/mechanism"
+	"path/filepath"
+	"strings"
+)
+
+type ingest struct {
+	uri         uri.Uri
+	location    location.Location
+	controller  *container.Controller
+	mechanism   mechanism.Mechanism
+	options     mechanism.BringOnlineOptions
+	extractor   extract.Extractor
+	containerId string
+	query       func(text string)
+}
+
+func (i *ingest) IsExtractionNeeded() bool {
+	i.extractor = extract.NewExtractor(i.uri.FileExtension(), i.controller)
+	if i.extractor == nil {
+		return false
+	} else {
+		return true
+	}
+}
+
+func (i *ingest) IsRemoteUrl() bool {
+	return !i.location.IsLocal()
+}
+
+func (i *ingest) UrlFilename() string {
+	return i.uri.Filename()
+}
+
+func (i *ingest) IsValidScheme() bool {
+	for _, s := range i.location.ValidSchemes() {
+		if s == i.uri.Scheme() {
+			return true
+		}
+	}
+	return false
+}
+
+func (i *ingest) CopyToContainer(containerId string) {
+	destFolder := "/var/opt/mssql/backup"
+	if i.mechanism != nil {
+		destFolder = i.mechanism.CopyToLocation()
+	}
+	if i.location == nil {
+		panic("location is nil, did you call NewIngest()?")
+	}
+
+	i.containerId = containerId
+	i.location.CopyToContainer(containerId, destFolder)
+	i.options.Filename = i.uri.Filename()
+
+	if i.options.Filename == "" {
+		panic("filename is empty")
+	}
+}
+
+func (i *ingest) Extract() {
+	if i.extractor == nil {
+		panic("extractor is nil")
+	}
+
+	if !i.extractor.IsInstalled(i.containerId) {
+		i.extractor.Install()
+	}
+
+	i.options.Filename, i.options.LdfFilename =
+		i.extractor.Extract(i.uri.Filename(), "/var/opt/mssql/data")
+
+	if i.mechanism == nil {
+		ext := strings.TrimLeft(filepath.Ext(i.options.Filename), ".")
+		i.mechanism = mechanism.NewMechanismByFileExt(ext, i.controller)
+	}
+}
+
+func (i *ingest) BringOnline(query func(string), username string, password string) {
+	if i.options.Filename == "" {
+		panic("filename is empty, did you call CopyToContainer()?")
+	}
+
+	i.query = query
+	i.options.Username = username
+	i.options.Password = password
+	i.mechanism.BringOnline(i.uri.GetDbNameAsIdentifier(), i.containerId, i.query, i.options)
+	i.setDefaultDatabase(username)
+}
+
+func (i *ingest) setDefaultDatabase(username string) {
+	if i.query == nil {
+		panic("query is nil, did you call BringOnline()?")
+	}
+
+	alterDefaultDb := fmt.Sprintf(
+		"ALTER LOGIN [%s] WITH DEFAULT_DATABASE = [%s]",
+		username,
```

Comment:
>username, [](http://example.com/codeflow?start=2&length=9)

user names need to be bracket escaped as well.
Fundamentally, we need a package with the same helper functions SMO has for safe scripting. See [https://github.com/microsoft/sqlmanagementobjects/blob/54e95cd095e232e43e711e65a83d55514ddc5e0e/src/Microsoft/SqlServer/Management/Smo/SqlSmoObject.cs#L329](https://github.com/microsoft/sqlmanagementobjects/blob/54e95cd095e232e43e711e65a83d55514ddc5e0e/src/Microsoft/SqlServer/Management/Smo/SqlSmoObject.cs#L329)